src/storage/store/column_chunk.cpp

```diff
@@ -0,0 +1,343 @@
+        }
+    } else {
+        for (auto i = 0u; i < numValuesToAppend; i++) {
+            auto posInChunk = startPosInChunk + i;
+            auto posInList = fixedSizedListArray->offset() + i;
+            memcpy(buffer.get() + getOffsetInBuffer(posInChunk),
+                valuesInList + posInList * numBytesPerValue, numBytesPerValue);
+        }
+    }
+}
+
+template<typename T>
+void ColumnChunk::templateCopyValuesAsString(
+    arrow::Array* array, offset_t startPosInChunk, uint32_t numValuesToAppend) {
+    auto stringArray = (arrow::StringArray*)array;
+    auto arrayData = stringArray->data();
+    if (arrayData->MayHaveNulls()) {
+        for (auto i = 0u; i < numValuesToAppend; i++) {
+            auto posInChunk = startPosInChunk + i;
+            if (arrayData->IsNull(i)) {
+                nullChunk->setNull(posInChunk, true);
+                continue;
+            }
+            auto value = stringArray->GetView(i);
+            setValueFromString<T>(value.data(), value.length(), posInChunk);
+        }
+    } else {
+        for (auto i = 0u; i < numValuesToAppend; i++) {
+            auto posInChunk = startPosInChunk + i;
+            auto value = stringArray->GetView(i);
+            setValueFromString<T>(value.data(), value.length(), posInChunk);
+        }
+    }
+}
+
+common::page_idx_t ColumnChunk::getNumPages() const {
+    auto numPagesToFlush = getNumPagesForBuffer();
+    if (nullChunk) {
+        numPagesToFlush += nullChunk->getNumPages();
+    }
+    for (auto& child : childrenChunks) {
+        numPagesToFlush += child->getNumPages();
+    }
+    return numPagesToFlush;
+}
+
+page_idx_t ColumnChunk::flushBuffer(
+    BMFileHandle* nodeGroupsDataFH, common::page_idx_t startPageIdx) {
+    // Flush main buffer.
+    FileUtils::writeToFile(nodeGroupsDataFH->getFileInfo(), buffer.get(), numBytes,
+        startPageIdx * BufferPoolConstants::PAGE_4KB_SIZE);
+    return getNumPagesForBuffer();
+}
+
+uint32_t ColumnChunk::getDataTypeSizeInChunk(common::LogicalType& dataType) {
+    switch (dataType.getLogicalTypeID()) {
+    case LogicalTypeID::STRUCT: {
+        return 0;
+    }
+    case LogicalTypeID::STRING: {
+        return sizeof(ku_string_t);
+    }
+    case LogicalTypeID::VAR_LIST: {
+        return sizeof(ku_list_t);
+    }
+    case LogicalTypeID::INTERNAL_ID: {
+        return sizeof(offset_t);
+    }
+    default: {
+        return StorageUtils::getDataTypeSize(dataType);
+    }
+    }
+}
+
+void FixedListColumnChunk::appendColumnChunk(kuzu::storage::ColumnChunk* other,
+    common::offset_t startPosInOtherChunk, common::offset_t startPosInChunk,
+    uint32_t numValuesToAppend) {
+    auto otherChunk = (FixedListColumnChunk*)other;
+    if (nullChunk) {
+        nullChunk->appendColumnChunk(
+            otherChunk->nullChunk.get(), startPosInOtherChunk, startPosInChunk, numValuesToAppend);
+    }
+    // TODO: This can be optimized to not copy one by one.
```

Comment:
Let's keep track of these TODOs somehow. Do you mind opening an issue about "Copy optimizations" and list these. We should assign it to someone and keep track of how much we improve.

Reply:
See https://github.com/kuzudb/kuzu/issues/1873.